crates/topos-tce-synchronizer/src/lib.rs

```diff
@@ -151,41 +158,62 @@ impl GrpcSynchronizerService for SynchronizerService {
                 return Err(Status::invalid_argument("Invalid checkpoint"));
             }
             Ok(value) => value,
         };
 
-        let diff = if let Ok(diff) = self.validator_store.get_checkpoint_diff(res) {
-            diff.into_iter()
-                .map(|(key, value)| {
-                    let v: Vec<_> = value
-                        .into_iter()
-                        .map(|v| ProofOfDelivery {
-                            delivery_position: Some(SourceStreamPosition {
-                                source_subnet_id: Some(v.delivery_position.subnet_id.into()),
-                                position: *v.delivery_position.position,
-                                certificate_id: Some(v.certificate_id.into()),
-                            }),
-                            readies: v
-                                .readies
-                                .into_iter()
-                                .map(|(ready, signature)| SignedReady { ready, signature })
-                                .collect(),
-                            threshold: v.threshold,
-                        })
-                        .collect();
-                    CheckpointMapFieldEntry {
-                        key: key.to_string(),
-                        value: v,
-                    }
-                })
-                .collect()
-        } else {
-            Vec::new()
+        debug!("Request {} contains {} proof_of_delivery", id, res.len());
+        trace!("Request {} contains {:?}", id, res);
+        let diff = match self.validator_store.get_checkpoint_diff(&res) {
+            Ok(diff) => {
+                debug!(
+                    "Fetched checkpoint diff from storage for request {}, got {:?}",
+                    id, diff
+                );
+                diff.into_iter()
+                    .map(|(key, value)| {
+                        let v: Vec<_> = value
+                            .into_iter()
+                            .map(|v| ProofOfDelivery {
+                                delivery_position: Some(SourceStreamPosition {
+                                    source_subnet_id: Some(v.delivery_position.subnet_id.into()),
+                                    position: *v.delivery_position.position,
+                                    certificate_id: Some(v.certificate_id.into()),
+                                }),
+                                readies: v
+                                    .readies
+                                    .into_iter()
+                                    .map(|(ready, signature)| SignedReady { ready, signature })
+                                    .collect(),
+                                threshold: v.threshold,
+                            })
+                            .collect();
+                        CheckpointMapFieldEntry {
+                            key: key.to_string(),
+                            value: v,
+                        }
```

Comment:
This looks like we're cloning the `ProofOfDelivery`s from the `diff` in order to create the `CheckpointMapFieldEntry`. Can't we just use `clone()`? I must be missing something. Is it the `source_subnet_id` and `delivery_position` the must be plucked out individually?

Reply:
This part is building the gRPC response based on the result of the query on the `store`.